### PR TITLE
Virtfs

### PR DESCRIPTION
**Pull request by hammerg**

### CODE REVIEW

File: viofs/pci/isrdpc.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	if (vq != NULL)
+	{
+		VirtFsReadFromQueue(context, vq, vq_lock);
+	}
+	else
+	{
+		for (i = 0; i < context->RequestQueues; i++)
+		{
+			vq = context->VirtQueues[i];
+            vq_lock = context->VirtQueueLocks[i];
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/isrdpc.c]: Few tabs sneaked it. I hope I handled all of them.

File: viofs/VirtFS Package/VirtFS Package.vcxproj [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Vista Release|Win32">
+      <Configuration>Vista Release</Configuration>
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Win8 Release|x64">
+      <Configuration>Win8 Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Win7 Release|x64">
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/VirtFS Package/VirtFS Package.vcxproj]: Only Win 8 and 10 for start. I don't know how these configurations reappeared. Will be removed.

File: viofs/pci/power.c [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	{
+		TraceEvents(TRACE_LEVEL_ERROR, DBG_POWER,
+			"Failed to allocate request queues");
+
+		status = STATUS_INSUFFICIENT_RESOURCES;
+	}
+
+    if (NT_SUCCESS(status))
+    {
+        context->VirtQueueLocks = ExAllocatePoolWithTag(NonPagedPool,
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: I'm using POOL_NX_OPTIN=1 so NonPagedPool value should be NonPagedPoolNx for Win8 and above.

File: viofs/pci/power.c [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    PDEVICE_CONTEXT context = GetDeviceContext(Device);
+
+    UNREFERENCED_PARAMETER(TargetState);
+
+    TraceEvents(TRACE_LEVEL_VERBOSE, DBG_POWER, "--> %!FUNC! Device: %p",
+        Device);
+
+    PAGED_CODE();
+
+    VirtIOWdfDestroyQueues(&context->VDevice);
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: The VirtIOWdfDestroyQueues do a device reset (status = 0) before deleting the queues.

AFAIK WDF handles the synchronizing issues so D0Exit is called after the interrupts are disabled.

File: viofs/pci/power.c [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+	VirtIOWdfDeviceGet(&context->VDevice,
+		FIELD_OFFSET(VIRTIO_FS_CONFIG, RequestQueues),
+		&context->RequestQueues,
+		sizeof(context->RequestQueues));
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_POWER,
+		"Request queues: %d", context->RequestQueues);
+
+	context->VirtQueues = ExAllocatePoolWithTag(NonPagedPool,
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: The DMA API will have to wait a bit. I'm still learning how it should be used in order to gain most of the virtio-fs features.

As for using ExAllocatePoolWithTag. When talking to the device, I'm sending a SG list build from the allocation MDL description. So the case of non-continuous memory allocation is handled.

File: viofs/pci/power.c [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+	if (!NT_SUCCESS(status))
+    {
+        TraceEvents(TRACE_LEVEL_ERROR, DBG_POWER,
+			"VirtIOWdfInitialize failed with %!STATUS!", status);
+    }
+
+	HostFeatures = VirtIOWdfGetDeviceFeatures(&context->VDevice);
+
+	if (virtio_is_feature_enabled(HostFeatures, VIRTIO_F_VERSION_1))
```
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: I'm okay with the current implementation.

File: viofs/pci/power.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ybendito) [reply viofs/pci/power.c]: This SG list is created without PCI SG interface, so these physicall addresses might be incorrect.

File: viofs/pci/power.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (YanVugenfirer) [reply viofs/pci/power.c]: But we don't have "real" interrupt disable in virtio. So DPC can still process the queue. Worth taking a second look.

File: viofs/pci/power.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (YanVugenfirer) [reply viofs/pci/power.c]: ExAllocatePoolWithTag - this is a bad practice. Let's stop doing it.

[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: So it is a good thing that I'm using MmAllocatePagesForMdlEx and not ExAllocatePoolWithTag.

[the remark this answers is not in the document]
Comment (ybendito) [reply viofs/pci/power.c]: MmAllocatePagesForMdlEx is also wrong as it does not take in account the device DMA capabilities. These addresses are good only for CPU access and for system without IOMMU.  Only class-driver allocators and  WDF DMA API do. If the driver (temporary) does not allocate the memory correctly it should prevent setting of VIRTIO_F_IOMMU_PLATFORM by explicit request to turn this feature off by VirtIOWdfSetDriverFeatures.

File: viofs/pci/power.c [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+	{
+		virtio_feature_enable(GuestFeatures, VIRTIO_F_VERSION_1);
+	}
+	if (virtio_is_feature_enabled(HostFeatures, VIRTIO_F_ANY_LAYOUT))
+	{
+		virtio_feature_enable(GuestFeatures, VIRTIO_F_ANY_LAYOUT);
+	}
+
```
Comment (ybendito) [inline viofs/pci/power.c]: The driver does not need to set explicitly VIRTIO_F_VERSION_1 and VIRTIO_F_ANY_LAYOUT, the WDF library does this when needed.

File: viofs/pci/power.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: For now just I'll just unset the VIRTIO_F_IOMMU_PLATFORM feature flag.

File: viofs/pci/power.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (hammerg) [reply viofs/pci/power.c]: After a second look, I think it is okay but I'll wait for bug reports ;-).

**Pull request closed without merging**